File: java/client-apis/src/main/java/org/apache/rocketmq/client/apis/consumer/FilterExpression.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.rocketmq.client.apis.consumer;

import static com.google.common.base.Preconditions.checkNotNull;

import com.google.common.base.Objects;

/**
 * Filter expression is an efficient way to filter message for {@link SimpleConsumer} and {@link PushConsumer}.
 * The consumer who applied the filter expression only can receive the filtered messages.
 */
public class FilterExpression {
    private static final String TAG_EXPRESSION_SUB_ALL = "*";
    public static final FilterExpression SUB_ALL = new FilterExpression(TAG_EXPRESSION_SUB_ALL);

    private final String expression;
    private final FilterExpressionType filterExpressionType;

    public FilterExpression(String expression, FilterExpressionType filterExpressionType) {
        this.expression = checkNotNull(expression, "expression should not be null");
        this.filterExpressionType = checkNotNull(filterExpressionType, "filterExpressionType should not be null");
    }

    /**
     * If the {@link FilterExpressionType} is not specified, the type is {@link FilterExpressionType#TAG}.
     *
     * @param expression tag filter expression.
     */
    public FilterExpression(String expression) {
        this(expression, FilterExpressionType.TAG);
    }

    /**
     * Default constructor, which means that messages are not filtered.
     */
    public FilterExpression() {
        this(TAG_EXPRESSION_SUB_ALL);
    }

    public String getExpression() {
        return expression;
    }

    public FilterExpressionType getFilterExpressionType() {
        return filterExpressionType;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        FilterExpression that = (FilterExpression) o;
        return expression.equals(that.expression) && filterExpressionType == that.filterExpressionType;
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(expression, filterExpressionType);
    }
}

File: php/Consumer/FilterExpression.php
<?php
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Apache\Rocketmq\Consumer;

/**
 * Filter expression class
 * 
 * Used to filter messages based on tags or SQL92 expressions
 */
class FilterExpression {
    /**
     * @var string Filter expression
     */
    private $expression;
    
    /**
     * @var FilterExpressionType Filter expression type
     */
    private $type;
    
    /**
     * Constructor
     * 
     * @param string $expression Filter expression
     * @param FilterExpressionType $type Filter expression type
     */
    public function __construct(string $expression, FilterExpressionType $type) {
        $this->expression = $expression;
        $this->type = $type;
    }
    
    /**
     * Get filter expression
     * 
     * @return string Filter expression
     */
    public function getExpression(): string {
        return $this->expression;
    }
    
    /**
     * Get filter expression type
     * 
     * @return FilterExpressionType Filter expression type
     */
    public function getType(): FilterExpressionType {
        return $this->type;
    }
    
    /**
     * Get string representation
     * 
     * @return string
     */
    public function __toString(): string {
        return sprintf("FilterExpression[expression=%s, type=%s]", $this->expression, $this->type);
    }
}
